examples_tests/22.RaytracedAO/raytraceCommon.glsl

```diff
@@ -271,10 +347,96 @@ nbl_glsl_MC_quot_pdf_aov_t gen_sample_ray(
 	in nbl_glsl_MC_instr_stream_t rnps
 )
 {
-	vec3 rand = rand3d(scramble_key,int(sampleID),int(depth));
+	mat2x3 rand = rand6d(scramble_key,int(sampleID),int(depth));
+
+	// (1) BXDF Sample and Weight
+	nbl_glsl_LightSample bxdfSample;
+	nbl_glsl_MC_quot_pdf_aov_t bxdfCosThroughput = nbl_glsl_MC_runGenerateAndRemainderStream(precomp,gcs,rnps,rand[0],bxdfSample);
+	float bxdfWeight = 0;
+
+	float p_bxdf_bxdf = bxdfCosThroughput.pdf; // BxDF PDF evaluated with BxDF sample (returned from 
+	float p_env_bxdf = Envmap_regularized_deferred_pdf(bxdfSample.L); // Envmap PDF evaluated with BxDF sample (returned by manual tap of the envmap PDF texture)
+
+	float p_env_env = 0.0f; // Envmap PDF evaluated with Envmap sample (returned from envmap importance sampling)
+	float p_bxdf_env = 0.0f; // BXDF evaluated with Envmap sample (returned from envmap importance sampling)
+
+	// (2) Envmap Sample and Weight
+	nbl_glsl_LightSample envmapSample;
+	nbl_glsl_MC_quot_pdf_aov_t envmapSampleThroughput;
+	{
+		nbl_glsl_MC_setCurrInteraction(precomp);
+
+		Envmap_generateRegularizedSample_and_pdf(/*out*/p_env_env, /*out*/ envmapSample, currInteraction.inner, rand[1].xy);
+
+		nbl_glsl_MC_microfacet_t microfacet;
+		microfacet.inner = nbl_glsl_calcAnisotropicMicrofacetCache(currInteraction.inner, envmapSample);
+		nbl_glsl_MC_finalizeMicrofacet(/*inout*/microfacet);
+
+		// bxdf eval_pdf_aov of light sample
+		const nbl_glsl_MC_eval_pdf_aov_t epa = nbl_bsdf_eval_and_pdf(precomp, rnps, 0xdeafbeefu, /*inout*/envmapSample, /*inout*/microfacet);
+  		envmapSampleThroughput.quotient = epa.value/epa.pdf;
+		envmapSampleThroughput.pdf = epa.pdf;
+		envmapSampleThroughput.aov = epa.aov;
```

Comment:
you can put all this (until line 433) in a `#ifndef ONLY_BXDF_SAMPLING`, since OpenGL compilers are dumb and don't optimize much (apparently example 42 runs much faster on Vulkan XD)

Reply:
Yea I thought compilers would optimize these away. 👍

Reply:
Done